amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/Scheduler.java

```diff
@@ -0,0 +1,702 @@
+            // Worker.run().
+            leaseCoordinator.stop();
+            workerStateChangeListener.onWorkerStateChange(WorkerStateChangeListener.WorkerState.SHUT_DOWN);
+        }
+    }
+
+    /**
+     * Perform final shutdown related tasks for the worker including shutting down worker owned executor services,
+     * threads, etc.
+     */
+    private void finalShutdown() {
+        log.info("Starting worker's final shutdown.");
+
+        if (executorService instanceof SchedulerCoordinatorFactory.SchedulerThreadPoolExecutor) {
+            // This should interrupt all active record processor tasks.
+            executorService.shutdownNow();
+        }
+        if (metricsFactory instanceof CloudWatchMetricsFactory) {
+            ((CloudWatchMetricsFactory) metricsFactory).shutdown();
+        }
+        shutdownComplete = true;
+    }
+
+    private List<ShardInfo> getShardInfoForAssignments() {
+        List<ShardInfo> assignedStreamShards = leaseCoordinator.getCurrentAssignments();
+        List<ShardInfo> prioritizedShards = shardPrioritization.prioritize(assignedStreamShards);
+
+        if ((prioritizedShards != null) && (!prioritizedShards.isEmpty())) {
+            if (slog.isInfoEnabled()) {
+                StringBuilder builder = new StringBuilder();
+                boolean firstItem = true;
+                for (ShardInfo shardInfo : prioritizedShards) {
+                    if (!firstItem) {
+                        builder.append(", ");
+                    }
+                    builder.append(shardInfo.shardId());
+                    firstItem = false;
+                }
+                slog.info("Current stream shard assignments: " + builder.toString());
+            }
+        } else {
+            slog.info("No activities assigned");
+        }
+
+        return prioritizedShards;
+    }
+
+    /**
+     * NOTE: This method is internal/private to the Worker class. It has package access solely for testing.
+     *
+     * @param shardInfo
+     *            Kinesis shard info
+     * @return ShardConsumer for the shard
+     */
+    ShardConsumer createOrGetShardConsumer(@NonNull final ShardInfo shardInfo,
+            @NonNull final ShardRecordProcessorFactory shardRecordProcessorFactory) {
+        ShardConsumer consumer = shardInfoShardConsumerMap.get(shardInfo);
+        // Instantiate a new consumer if we don't have one, or the one we
+        // had was from an earlier
+        // lease instance (and was shutdown). Don't need to create another
+        // one if the shard has been
+        // completely processed (shutdown reason terminate).
+        if ((consumer == null)
+                || (consumer.isShutdown() && consumer.shutdownReason().equals(ShutdownReason.LEASE_LOST))) {
+            consumer = buildConsumer(shardInfo, shardRecordProcessorFactory);
+            shardInfoShardConsumerMap.put(shardInfo, consumer);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()` and `put()` may not be thread-safe at lines: **569, 575, and 578**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread's `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and returns that reference to its caller at line: **581**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.